Fixes: #35. Add `date` and `expires` to signature

ioggstream opened the pull request:
## This PR

Adds  `date` and `expires` fields to the Signature.

ioggstream committed:
Fixes: #35. Add `date` and `expires` to signature

Add `date` and `expires` to the Signature.

msporny commented on index.xml:
```diff
@@ -175,6 +175,23 @@ of keys and assignment of `keyId` is out of scope for this document.
      </t>
     </section>
 
+    <section anchor="date" title="date">
```
I suggest we use a signature parameter called `created`, date is too vague.

HTTP Signatures don't follow the typical digitally signed documents use cases (where there is an issued date and an expiration date). They are intended to be instantaneous signatures, useful at the time they're created plus or minus a delta that is decided by the server.

@dlongley also mentioned another problem, and that's in Javascript environments, where you don't have control over the Date header and would need to use this value. 

We need to mention that implementations MUST use `created` if it exists to determine the signature validity period and then fall back to Date if `created` is not provided, and SHOULD reject the signature if no other replay prevention mechanism is utilized.

The expression needs to support something better than unix timestamps, providing millisecond precision (and even further than that in time) - HTTP Signatures are used in financial trading systems where you need millisecond (or submillisecond precision). I suggest we use ISO8601 with microsecond precision, if available.

ioggstream replied:
Hi M, thanks for your feedback!

1- on milliseconds, it could be fine, I just took the fields from signed-exchanges and I think we should somewhat converge. Sub-millisecond usually fight with latencies and time-skew (see below). 

2- iiuc created/expires are the ws-security names. To me is fine.

ioggstream replied:
Using `date` could be useful in working around this: https://github.com/w3c-dvcg/http-signatures/issues/33

msporny replied:
> Sub-millisecond usually fight with latencies and time-skew

Yes, but they make great monotonically increasing nonces for things like financial trading systems.

liamdennehy replied:
I'd go with ``issued``, as it aligns with token-type systems like jwt (iat). This is after all the date the signature validity commenced, and is independent of the "date" of the payload. If it's important to the payload it should be encoded there. It was always problematic that we're using the default Date header, as it is even banned from requests under certain schemes and is meant to indicate the date of transmission, again separate from the document or signature date values.

msporny commented on index.xml:
```diff
@@ -175,6 +175,23 @@ of keys and assignment of `keyId` is out of scope for this document.
 
+    <section anchor="date" title="date">
+     <t>
+REQUIRED.  The `date` field is an integer representing a Unix time. Represent the
```
We should make this RECOMMENDED. Doing REQUIRED will break all existing implementations in the field.

ioggstream replied:
about MUST/SHOULD: let's discuss that *after* we found a suitable spec. A good solution could be to *version* the Signature header so that we can fix future security issues without breaking old stuff.

msporny commented on index.xml:
```diff
@@ -175,6 +175,23 @@ of keys and assignment of `keyId` is out of scope for this document.
+     </t>
+    </section>
+
+    <section anchor="expires" title="expires">
```
We don't need expires. We just need one time to be specified. It's up to the server to determine if that timestamp is valid for the purpose at hand.

ioggstream replied:
On instant signatures, see https://github.com/w3c-dvcg/http-signatures/issues/35#issuecomment-417238049

ioggstream replied:
It's optional now. Still I want a way for the client to control the expiration of the signature.

msporny commented on index.xml:
```diff
@@ -577,6 +594,7 @@ The signature header and signature would be generated as:
     <figure>
      <artwork>
 Signature: keyId="rsa-key-1",algorithm="rsa-sha256",
+date=1402170695, expires=1402170699,
```
Suggested instead: `created="2018-08-20T11:27:41.799638807Z"`

ioggstream replied:
I first proposed an internal standard based on RFC3339 but a couple of security guys at Google suggested to use timestamps to avoid parsing issues. RFC3339 and ISO should theoretically manage timezone offset too, which complicates validation.

msporny reviewed (changes requested):
Some thoughts on what could be changed to pull the PR in.

ioggstream committed:
`date` and `expires` are optional. 

## This PR

  - allows `date` and `expires` as optional values
  - `Date` header for requests (eg in IoT) can be worked-around via the `date` field (same name, but in header).
  - subseconds precision is allowed via floating values
  - still in unix timestamp because parsing RFC3339 is complex (timezones, type hinting, ...) while you can always use a float.

liamdennehy commented:
I still don't understand what value this has in the signature header - if it isn't included in the signature itself, what value does it have as a statement? It's trivially altered, so cannot be relied on.

ioggstream commented:
Hi @liamdennehy! 

> I still don't understand what value this has in the signature header - if it isn't included in the signature itself, what value does it have as a statement? It's trivially altered, so cannot be relied on.

I marked this PR as [WIP] because it's not complete and serve to discuss the changes that will lead to a complete - non [WIP] PR.

I will add that soon, anyway :)

ioggstream committed:
Replace  to `float` with `decimal notation`

ioggstream commented on index.xml:
```diff
@@ -175,6 +175,27 @@ of keys and assignment of `keyId` is out of scope for this document.
+signature issuing timestamp. A signature generated after the current timestamp
+should not be processed. Using Unix timestamp avoids speeds up processing and
+avoid timezone management existing in RFC3339. Subsecond precision is allowed
+using decimal notation.
```
@liamdennehy replaced with `decimal notation`. Let me know if it's good for you. To me - considering network and reverse-proxy latency - integers are fine.

liamdennehy replied:
Not sure why I was tagged, but ok :smile:

ioggstream replied:
I received a mail from gh with you asking `which float spec?` ;)

liamdennehy replied:
Sorry, you did indeed :smile: I need to pay more attention

liamdennehy commented:
> Hi @liamdennehy!
> 
> > I still don't understand what value this has in the signature header - if it isn't included in the signature itself, what value does it have as a statement? It's trivially altered, so cannot be relied on.
> 
> I marked this PR as [WIP] because it's not complete and serve to discuss the changes that will lead to a complete - non [WIP] PR.
> 
> I will add that soon, anyway :)

So I propose we move these values to distinct headers (using the format proposed already to cater for sub-second precision):
```
Signature-Issued: 1542330444
Signature-Expires: 1542331344
```
Or, should expires be expressed as a delta from issued? This turns statements into "now + 900 seconds", maybe simplifying.

This also preserves signature date as distinct from message date, which may have different meanings.

ioggstream commented:
- timestamp or delta: both https://github.com/WICG/webpackage/blob/master/draft-yasskin-httpbis-origin-signed-exchanges-impl.md#examples--example-signature-header and https://tools.ietf.org/html/rfc7519#section-4.1.4 use full timestamp. I'd follow that road.

- about the missing `Date` headers in requests: the ability to specify the `date` directly in the `Signature` allow us to not necessarily mess with headers not governed by us.

- about headers: though I wish for a progressive convergence with  https://github.com/WICG/webpackage/blob/master/draft-yasskin-httpbis-origin-signed-exchanges-impl.md#examples--example-signature-header in a way that we'll eventually end in a single `Signature` header for httpbis, I think that it's an improvement.

liamdennehy commented:
> * about headers: though I wish for a progressive convergence with  https://github.com/WICG/webpackage/blob/master/draft-yasskin-httpbis-origin-signed-exchanges-impl.md#examples--example-signature-header in a way that we'll eventually end in a single `Signature` header for httpbis, I think that it's an improvement.

If I understand httpbis properly, it merely adds a second (or more?) identity for authentication. It does not produce digitally signed messages, and since TLS is still based on symmetric ciphers it cannot provide non-repudiation. I think this spec still stands on its' own in that regard.

ioggstream commented:
iiuc the `Signature` header in signed-exchanges proof to a 3rd party that:

  - the requester is provided with proof of delivery (aka the signed response)

The keys used are ECC (asymmetric).

liamdennehy commented:
> iiuc the `Signature` header in signed-exchanges proof to a 3rd party that:
> 
>     * the requester is provided with proof of delivery (aka the signed response)
> 
> 
> The keys used are ECC (asymmetric).

You're right, I got that proposal confused with another. Wow, that's one complex proposal. If this scheme is JSON, that one's XML: Waaay too focused on implementation with cross-signing, caching and all kinds of other goodies. I just wanna sign my HTTP Request & Response plz. :stuck_out_tongue:

ioggstream committed:
Use the (date) and (expires) special headers.

Define optional special headers for (date) and (expires).

ioggstream commented on index.xml:
```diff
@@ -303,6 +324,26 @@ x-example: Example header with some whitespace.
    </t>
 
+        <t>
+If you want to add an optiona date or expires to the signature, you
```
@liamdennehy do you like it? This should work with `fetch` too.

ioggstream replied:
@dlongley this should address your comment when `Date` is either:

- forbidden
- not mandatory

dlongley replied:
Yeah, I like this approach. We might want `created` instead of `date` though (bikeshed).

msporny replied:
Since this changes the signature algorithm a bit, I suggest it goes out in the new spec under a new "algorithm" mechanism... that is, rsa-sha1 (or whatever ancient thing we're using right now) should be changed to what we're doing in the Linked Data Proof/Signature we're doing at W3C... that is, we give the signing string construction algorithm a yearly identifier so we don't break future upgradability. rsa-sha1 will remain a "legacy" mechanism that we deprecate over time. 

Agree with where this PR is going... nicely done @ioggstream!

ioggstream replied:
@dlongley I inherited the `date` and `expires` from https://tools.ietf.org/id/draft-yasskin-http-origin-signed-responses-05.html#rfc.section.3.1 (which are probably going to register `Signature` into iana).

@msporny thanks for your feedback! It's a long time :)

1- I think we should plan for a versioned release of http-signatures. If `Signatures` will be registered by https://github.com/WICG/webpackage/blob/master/draft-yasskin-http-origin-signed-responses.md#iana-considerations   we could move to a new header with a more precise semantic. 

2- In the simple case, we can just add a new optional parameter containing the http-signature version, like it's done in other specs. I found this interesting reading on the subject https://tools.ietf.org/id/draft-thomson-postel-was-wrong-03.html.

msporny commented on index.xml:
```diff
@@ -603,6 +645,7 @@ be generated as:
     <figure>
      <artwork>
 Signature: keyId="rsa-key-1",algorithm="rsa-sha256",
```
I suggest that this is a breaking change and so we should create a new "algorithm" for this... something like "rsasig-2019", and then define that to use something like "rsa + sha256". We should probably also define an Ed25519 equivalent "ed25519sig-2019", which uses eddsa + sha256. This enables us to keep backwards compatibility with rsa-sha256 (because the old implementations will silently fail w/ the new optional parameters.

ioggstream replied:
The question is whether we want to version algorithms or signatures, namely the string to sign. See my new proposal #53 

I'd stick with the current algos naming, so that people can easily reference them. And add another, mandatory signed field named `v` usable to specify the version.

This allows our users to deprecate older versions of the spec using a standard, versioned parser. Eg.

```
Signature: v=1.1; ... 

# I require (2> v > 1.5) 
```

Instead of having them to search in their algorithm's repo.

dlongley replied:
IIRC, we want to discourage the use of the `algorithm` parameter in favor of getting this information from the key instead: https://github.com/w3c-dvcg/http-signatures/blob/43577c3e6debd30ba3fc56c3bfe4c783f3a059f9/index.xml#L218-L228

Also, the `algorithm` parameter is primarily about the crypto rather than building the signature string, as @ioggstream says. So it sounds like we want some other field here (version or similar).

msporny reviewed (changes requested):
This is a backwards incompatible change and will create horrible security issues if we don't upgrade the "algorithm" parameter in the process. We should do that in the next release. All that to say, this should go in, but in a way that doesn't break the existing deployments.

msporny reviewed (approved):
This PR still needs a bit of work, rather than wait for it to happen here, I'm going to pull it in and do the rest and make a new release. At the most basic level, we'll add these new features, but do it in a way that 1) doesn't stomp over the special HTTP pseudo-headers, 2) that signals that there is a new canonicalization algorithm, and 3) doesn't break backwards compatibility.
